packages/protocol/tipmanager/tipmanager.go

```diff
@@ -216,19 +174,171 @@ func (t *TipManager) selectTips(count int) (parents models.BlockIDs) {
+func (t *TipManager) PromoteFutureTips(cm *commitment.Commitment) {
+	t.evictionMutex.RLock()
+	defer func() {
+		t.evictionMutex.RUnlock()
+		t.Evict(cm.Index())
+	}()
+
+	if futureEpochTips := t.futureTips.Get(cm.Index()); futureEpochTips != nil {
+		if tipsForCommitment, exists := futureEpochTips.Get(cm.ID()); exists {
+			tipsToPromote := make(map[models.BlockID]*scheduler.Block)
+			tipsForCommitment.ForEach(func(blockID models.BlockID, tip *scheduler.Block) bool {
+				tipsToPromote[blockID] = tip
+				return true
+			})
+
+			tipsToNotPromote := set.NewAdvancedSet[models.BlockID]()
+			for _, tip := range tipsToPromote {
+				for _, parentID := range tip.Parents() {
+					if _, exists := tipsToPromote[parentID]; exists {
+						// Do not add this tips parent because this is not a tip anymore
+						tipsToNotPromote.Add(parentID)
+					}
+				}
+				// Remove all tips referenced by all future tips valid for this commitment
+				t.RemoveStrongParents(tip.ModelsBlock)
+			}
+
+			for tipID, tip := range tipsToPromote {
+				if !tipsToNotPromote.Has(tipID) {
+					t.addTip(tip)
+				}
+			}
+		}
+	}
+}
+
+// Evict removes all parked tips that belong to an evicted epoch.
+func (t *TipManager) Evict(index epoch.Index) {
+	t.evictionMutex.Lock()
+	defer t.evictionMutex.Unlock()
+
+	t.futureTips.Evict(index)
+}
+
 // endregion ///////////////////////////////////////////////////////////////////////////////////////////////////////////
 
 // region TSC to prevent lazy tips /////////////////////////////////////////////////////////////////////////////////////
 
+func (t *TipManager) addTip(block *scheduler.Block) (added bool) {
+	if !t.tips.Has(block.ID()) {
+		t.tips.Set(block.ID(), block)
+		// t.tipsConflictTracker.AddTip(block)
+		t.Events.TipAdded.Trigger(block)
+
+		// skip removing tips if a width is set -> allows to artificially create a wide Tangle.
+		if t.TipCount() <= t.optsWidth {
+			return true
+		}
+
+		// a tip loses its tip status if it is referenced by another block
+		t.RemoveStrongParents(block.ModelsBlock)
+
+		return true
+	}
+
+	return false
+}
+
+// checkMonotonicity returns true if the block has any accepted or scheduled child.
+func (t *TipManager) checkMonotonicity(block *scheduler.Block) (anyScheduledOrAccepted bool) {
+	for _, child := range block.Children() {
+		if child.IsOrphaned() {
+			continue
+		}
+
+		if t.blockAcceptanceGadget.IsBlockAccepted(child.ID()) {
+			return true
+		}
+
+		if childBlock, exists := t.schedulerBlockRetrieverFunc(child.ID()); exists {
+			if childBlock.IsScheduled() {
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+// isStaleCommitment returns true if the block belongs to a commitment that is too far back in the past.
+func (t *TipManager) isStaleCommitment(block *scheduler.Block) (isOld bool) {
+	return block.Commitment().Index() < (t.engine.Storage.Settings.LatestCommitment().Index() - 1).Max(0)
```

Comment:
Shouldn't that also use the epochDelay as defined in `isRecentCommitment`?  That way it could call `isStaleCommitment` instead of repeating the comparison.
```
epochDelay := epoch.Index(int64(t.optsTimeSinceConfirmationThreshold.Seconds()) / epoch.Duration)
```

Reply:
Yes indeed, I can change the order of the checks in `AddTip` and reuse the same check method.